bip-ccv.mediawiki

```diff
@@ -0,0 +1,416 @@
+
+The following code is executed every time the <code>OP_CHECKCONTRACTVERIFY</code> opcode is encountered during the evalutation of a taproot Script spend. <code>this_input_index</code>, <code>this_input_internal_key</code> and <code>this_input_taptree</code> are the index, taproot internal key and taproot Merkle root of the current input.
+
+<code>BIP341_NUMS_KEY</code> is the x-only provably unspendable key <code>50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0</code> defined in [[bip-341.mediawiki|BIP-341]].
+
+<source lang="python">
+  if flags < CCV_FLAG_CHECK_INPUT or flags > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT:
+    return success()  # undefined flags are OP_SUCCESS
+
+  if index == -1:
+    index = this_input_index
+
+  if flags == CCV_FLAG_CHECK_INPUT:
+    if index < 0 or index >= n_inputs:
+      return fail()  # input index out of bounds
+
+    target_script = inputs[index].scriptPubKey
+  else:
+    if index < 0 or index >= n_outputs:
+      return fail()  # output index out of bounds
+
+    target_script = outputs[index].scriptPubKey
+
+  if taptree == <-1>:
+    taptree = this_input_taptree
+
+  if pk == <0>:
+    naked_key = BIP341_NUMS_KEY
+  elif pk == <-1>:
+    naked_key = this_input_internal_key
+  elif len(pk) == 32:
+    naked_key = pk
+  else:
+    return fail()
+
+  # Verify the target contract data and program
+  _, internal_key = tweak_embed_data(naked_key, data)
+
+  if len(taptree) != 0:
+    if len(taptree) != 32:
+      return fail()
+
+    _, final_key = taproot_tweak_pubkey(internal_key, taptree)
+  else:
+    final_key = internal_key
+
+  if target_script != P2TR(final_key):
+    return fail()
+
+  # Amount checks
+
+  if flags == 0:
+    # default amount semantic
+    if output_checked_deduct[index]:
+      return fail()
+
+    output_min_amount[index] += residual_input_amount
+    residual_input_amount = 0
+
+    if outputs[index].amount < output_min_amount[index]:
+      return fail()
+
+    output_checked_default[index] = True
+  elif flags == CCV_FLAG_DEDUCT_OUTPUT_AMOUNT:
+    # 'deduct' amount semantic
+    if residual_input_amount < outputs[index].amount:
+      return fail()
+
+    if output_checked_default[index] or output_checked_deduct[index]:
+      return fail()
+
+    residual_input_amount = residual_input_amount - outputs[index].amount
+    output_checked_deduct[index] = True
+
+  stack.pop(5)  # drop all 5 stack elements
+</source>
+
+==== sigops budget ====
+
+TODO
+
+== Policy changes ==
+
+TODO
+
+== Implementation ==
+
+A reference implementation is provided here:
+
+TODO
+
+== Examples ==
+
+This section documents some common Script fragments that use <code>OP_CHECKCONTRACTVERIFY</code> for various common choices of the parameters. Depending on the use case, some of the parameters might be passed via the witness stack.
+In these examples, <code><></code> (empty buffer) and <code>0</code> both refer to an empty stack element.
```

Comment:
Can you frame these examples in the context of an actual application?

Reply:
In 9bccf545930f9d4b8211f11d5bd93eac27b75695 I added one- or two-sentence explainers after each example (and deleted the example with just the _deduct_ mode, as the one right after already has a _deduct_ check followed by a _default_ one - which is the correct way of using it anyway).